baybe/searchspace/discrete.py

```diff
@@ -710,6 +713,45 @@ def parameter_cartesian_prod_to_df(
     return ret
 
 
+def parameter_cartesian_prod_polars(
+    parameters: Iterable[Parameter], constraints: Iterable[DiscreteConstraint]
+) -> pd.DataFrame:
+    """Create the Cartesian product of all parameter values.
+
+    Ignores continuous parameters.
+
+    Args:
+        parameters: List of parameter objects.
+        constraints: List of constraint objects.
+
+    Returns:
+        A dataframe containing all possible discrete parameter value combinations.
+    """
+    discrete_parameters = [p for p in parameters if isinstance(p, DiscreteParameter)]
+    if not discrete_parameters:
+        return pd.DataFrame()
+
+    param_frames = [pl.LazyFrame({p.name: p.values}) for p in discrete_parameters]
+    if len(param_frames) < 1:
+        return pd.DataFrame()
+
+    if len(param_frames) == 1:
+        return param_frames[0].collect().to_pandas()
+
+    res = param_frames[0]
+
+    for frame in param_frames[1:]:
+        # Join first
+        res = res.join(frame, how="cross")
+        # Filter constraints
+        for c in constraints:
+            pl_expr = c.get_polars_expression(res)
```

Comment:
why do the constraint applciation int his product utility? the old pandas cosntraint filtering is in tis own utlity, so another utility for polarsz makes sense to me

Reply:
You mean I move this part of the code to a new function like `_apply_polars_constraint()`?

```
for c in constraints:
            pl_expr = c.get_polars_expression(res)
            if pl_expr is not None:
                res = res.filter(pl_expr)
```

Reply:
yes and imo it also shouldnt be in the carteisan product function
here is conceptually how i imagine `.from_product` should work

```python
constraints_polar, constraints_pandas = _split_constraints(constraints) # splits them based on hasattr get_polars_expression

polars_df = cartesian_product(parameters)
polars_df_filtered = _apply_constraints_polars(constraints_polars)
pandas_df = polars_df.collect()
pandas_df_filtered = _apply_constraints_pandas(constraints_pandas)

return pandas_df_filtered
```

Or is there a numerical reason why the polars constraints need to be included already during the product creation?

Reply:
> Or is there a numerical reason why the polars constraints need to be included already during the product creation?

Based on a couple of experiments that I had, Polars cross join operation is not faster than Pandas. However, by applying constraints during join/product operation, we are able to reduce the number of parameters/inputs and hence improve the execution time.

Reply:
**Update 2:** In larger parameter spaces, filtering during product operation could be ~2x faster, e.g:
10 parameters, each 5 possible values 
Filtering **during** product execution time: 98.6 ms ± 58.9 ms per loop
Filtering **after** product execution time: 172 ms ± 51.8 ms

Reply:
Hi @rjavadi, can you comment on how you measured the times? I.e. which operations are covered exactly? Also: the  size of the space you used is rather small, can you crank it up by some considerable factor? Because as long as we are in the ms-range, we are not in the regime for which we even need the polars switch 🙃 Let's take something of at least a couple of seconds, otherwise it could simply be overhead that dominates...

Reply:
@AdrianSosic 
I added the details [here](https://github.com/emdgroup/baybe/discussions/267)